Radius support for AWS IRSA (#51)
radius-project/design-notes: cli/2024-06-04-aws-irsa-support.md

# Radius AWS Workload Identity Support

* **Status**: In Review
* **Author**: Nithya Subramanian (@nithyatsu)

## Overview

> A software workload such as a container-based application, service or script needs an identity to authenticate, access, and communicate with services that are distributed across different platforms and/or cloud providers. Radius uses AWS IAM (Identity and Access Management) credentials to deploy and access AWS resources today:

Reviewer: ```suggestion
A software workload such as a container-based application, service or script needs an identity to authenticate, access, and communicate with services that are distributed across different platforms and/or cloud providers. Radius uses static credentials (Access Key) to interact with AWS IAM (Identity and Access Management) so that we can deploy and access AWS resources today:
```

 
```
rad credential register aws --access-key-id <access-key-id> --secret-access-key <secret-access-key>
```
 
These IAM credentials, including the access key should be rotated regularly to reduce the chance of unauthorized access. A more secure option than using these credentials is to use IRSA.
 
IRSA (IAM Roles for Service Accounts) is used for authenticating applications running within Kubernetes pods. When we use IRSA, we associate an IAM role directly with a Kubernetes service-account, allowing the pods associated with the service-account to assume that role. IRSA does not require us to configure access key and  secret. It relies identity tokens (OIDC tokens) to authenticate with AWS services. These tokens have a default expiration (configurable). However, they automatically refresh when needed, transparent to the user.

The goal of the scenario is to enable infrastructure operators to configure IRSA support for the AWS provider in Radius to deploy and manage AWS resources.

> ## Terms and definitions

Reviewer 2: add roleARN, ODIC to the list of term definitions

> 

Reviewer: Any terms we should define here?

Reviewer: I think we should define some of the vocabulary used in the doc to describe IAM and it's features. eg: role ARN (I'm sure there are others)

Reviewer 2: e.g. ODIC

Author: added them.

## Objectives

> Issue Reference: https://github.com/radius-project/radius/issues/7618 

### Goals

* Radius users can configure AWS provider and enable Radius to use IRSA for authentication and deployment of application.
* IRSA can be configured via interactive experience.
* IRSA can be configured manually.
* Radius users can deploy and manage AWS resources without having to manage AWS access key id and secret

### Non-goals

* Azure Managed Identity support
* Azure Workload Identity support

> * Ability of radified applications to be able to use IRSA for authenticating AWS operations.

Reviewer 3: Is this a non-goal?

Reviewer 3: Is the keyword "operations" here?

> * Ability of radified applications to be able to use IRSA for authenticating AWS operations.

Reviewer 4: Could you explain this a little more? I'm not sure I understand the difference between this and the goal "Radius users can deploy and manage AWS resources without having to manage AWS access key id and secret"

Author: added some details.

### User scenarios

As a user, I should be able to configure Radius with IRSA for managing AWS resources.

## Design

### High Level Design for IRSA

> During installation, Radius should allow user to enable IRSA for AWS while conifguring AWS provider. Then the user should be able to enter the roleARN associated with desired policies which will be stored as the AWS credential.

Reviewer: Does the user have to associate an ARN during installation?

Author: no it does not. updated.

At a high level, the full flow to setup IRSA looks as below:

> 1. configure the cluster with oidc provider. have the role-arn configured with desired policies ready.

Reviewer 3: Can we have 2 OIDC providers in the same cluster? Asking because what if we have resources on Azure and also on AWS?

Author: same OIDC should  provide the authentication needed for both aws and azure.

> 2. take steps to allow  `rad-role-arn` and `eks cluster : namespace: service` to trust each other using aws portal.

Reviewer 5: What if the Kubernetes cluster isn't EKS?

Author: The steps to setup OIDC provider for that cluster and establish trust between AWS and the cluster will depend on the cluster. Radius would not need any additional changes. https://github.com/smalltown/aws-irsa-example/blob/master/README.md has a write up that details steps for self-hosted cluster.

3. use  `rad init` to configure radius with role-arn credential. (or use rad credential later on)

> 4. use `rad env update` to configure provider associated with the specific environment (`rad env update qa --aws-account-id <aws-account-id> --aws-region <aws-region>`). 

Reviewer 6: I think rad init should take care of updating the environment

> 5. during rad deploy, radius assumes the role that is configured in the credential. 

Reviewer 2: can the configured IRSA creds also be used to pull from private registries and repos on AWS? e.g. pulling bicep recipes from private registries

Author: I havent tried this out, but I think it should based on whether the role has the right policy.

> 

Reviewer: I think it would be helpful if we could describe the steps that need to be taken inside AWS vs inside Radius. Can you link the docs for AWS instead of trying to re-describe?

Author: I have rewritten this section.

Since oidc is configured for a cluster and not specific to a role-arn, the solution can be extended to work in multi-tenancy once we have multi credentials support. 

More details about each step is covered in [Detailed Design](#detailed-design)

> IRSA works the same way on EKS as well as non-EKS clusters.  

Reviewer: 👍 glad you called this out.

> IRSA works the same way on EKS as well as non-EKS clusters.  

Reviewer 7: What is the expected IAM policy for non-EKS clusters?

Author: the IAM policy should allow Radius to manage resources we support and follow the principle of least privilege as good practice. It would be the same on both eks and non-eks cluster.

Reviewer 7: I was trying to get more specific details about what the policy would look like similar to the example you added here https://github.com/radius-project/design-notes/blob/8aca2d8a3b550ed8f1035d0ce57c2034999ed4a0/cli/2024-06-04-aws-irsa-support.md?plain=1#L164, to understand it better. :) Feel free to share a link to AWS docs if it is already documented on EKS docs.

Author: The Principal looks generally similar to  
```
"Principal": {
          "Federated": "arn:aws:iam::${ACCOUNT_ID}:oidc-provider/${OIDC_PROVIDER}"
        },
```
This is applicable if we use a AWS OIDC provider, irrespective of the cluster.  If we want another OIDC provider, https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html covers steps for configuring that, and them using the IAM of that provider.

### Alternatives considered

#### EKS Pod Identity 

> EKS Pod Identity was introduced in 2022 as a simplified approach for applications running on EKS to retrieve credentials. It uses the new EKS pod identity instead of OIDC provider for identity. Because of this, users have convenient APIs that allow managing pod identity. 

Reviewer: > EKS pod identity instead of OIDC provider for identity

Can you explain what's different about these technologies?

I'm trying to square this with your later statement that:

> Radius should be however able to support this feature with minimal code changes, if any.

Author: I meant, both approaches allows Radius to assume a AWS role effectively, once the configuration on AWS side is complete. Since we are bringing in support for role-arn as credential, we would not need any code changes.

Ref. [AWS EKS Pod Identity](https://aws.amazon.com/blogs/containers/amazon-eks-pod-identity-a-new-way-for-applications-on-eks-to-obtain-iam-credentials/). 

Since the identity provider is tied to AWS, this solution cannot be used with Radius deployed on stand alone, aks or any other kubernetes cluster. 

Its good to note that since this approach uses its own service principal, the setup involves different identity providers and configuration on aws. Radius should be however able to support this feature with minimal code changes, if any.  

### Architecture Diagram

> 

Reviewer: Yes please :)

### Detailed Design

#### Proposed Option

Radius should support AWS IRSA.

#### Prerequisite

1. [Setup OIDC provider] (https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html) should be complete for the cluster. 

> 2. `rad-role` with desired policies should be created. Its trust relation should look similar to below. This establishes a trust relation between the annotated service-accounts in the specified namespace in specified cluster with the specified OIDC provider to be established. 

Reviewer: is `rad-role` significant? If the user creates this can they call it anything they want?

Author: yes. its just AWS role attached to right policies that will allow radius to manage AWS resources effectively.

> 2. `rad-role` with desired policies should be created. Its trust relation should look similar to below. This establishes a trust relation between the annotated service-accounts in the specified namespace in specified cluster with the specified OIDC provider to be established. 

Reviewer 2: what is `rad-role` in this context?

Author: renamed it to radius-role. It is the AWS IAM role that radius should assume, so that it can manage all its supported AWS operations.

> 2. `rad-role` with desired policies should be created. Its trust relation should look similar to below. This establishes a trust relation between the annotated service-accounts in the specified namespace in specified cluster with the specified OIDC provider to be established. 

Reviewer 7: nit: I'd suggest calling it `radius-role` in documentation/examples since explicit role names are usually preferred so that the purpose is clear from the name. Users can decide to name it however they want as Ryan called out above, but we should guide them towards right conventions in docs.

  ```
  {
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Federated": "arn:aws:iam::817312594854:oidc-provider/oidc.eks.us-west-2.amazonaws.com/id/67DDAC18D8C44CEDCF1C9719A8E9B866"
            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "StringEquals": {
                    "oidc.eks.us-west-2.amazonaws.com/id/67DDAC18D8C44CEDCF1C9719A8E9B866:sub": "system:serviceaccount:radius-system:ucp",
                    "oidc.eks.us-west-2.amazonaws.com/id/67DDAC18D8C44CEDCF1C9719A8E9B866:aud": "sts.amazonaws.com"
                }
            }
        },
    ]
}

>   ```

Reviewer: In this policy document, what are the things that are static and what are the things that will change from user-to-user? Can you point out where identifiers from this document need to match something in Radius?

Author: I will add the radius specific explanation and capture ucp + recipe rp details.

  Note: 67DDAC18D8C44CEDCF1C9719A8E9B866 is the specific eks cluster instance.
  The aud is the intended recipient of the token.
  The sub is the entity that the token represents.

#### Radius support

**installation**

> User can enable IRSA for AWS by by choosing "IRSA" option by choosing IRSA during rad init and providing a RoleARN (`arn:aws:iam::123456789012:role/rad-role`). This stores the roleARN as credential. It also mounts the service-account token to ucp and rp pods as secret.

Reviewer 3: So basically, the steps are:

1. Create an EKS Cluster (only EKS) with OIDC
2. During `rad init` enable IRSA
3. Update the Radius environment (Do we need this for sure?)

Downsides (my opinion):
1. A lot of steps
2. Can only work on EKS (which may be okay with AWS users)
3. Can't (?) deploy resources to other clouds

> User can enable IRSA for AWS by by choosing "IRSA" option by choosing IRSA during rad init and providing a RoleARN (`arn:aws:iam::123456789012:role/rad-role`). This stores the roleARN as credential. It also mounts the service-account token to ucp and rp pods as secret.

Reviewer: Do we need to know the role at installation time? Or is the role associated with a credential?

Author: updated.

```
    Containers:
      ucp:
        Container ID:   docker://faa5af929f91e053b04b023e05c4fa6363e87688df265443d9f377258fc3fd04
        Image:          ghcr.io/radius-project/ucpd:latest
        Image ID:       docker-pullable://ghcr.io/radius-project/ucpd@sha256:991ed703a3260a19535690242cef4d7e4d076038b016443d9106f2d76d57b0da
        Ports:          9443/TCP, 9090/TCP
        Host Ports:     0/TCP, 0/TCP
        :
        Environment:
          UCP_CONFIG:                   /etc/config/ucp-config.yaml
          BASE_PATH:                    /apis/api.ucp.dev/v1alpha3
          TLS_CERT_DIR:                 /var/tls/cert
          PORT:                         9443
        Mounts:
          /etc/config from config-volume (rw)
          /var/run/secrets/eks.amazonaws.com/serviceaccount from aws-iam-token (ro)
          /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-d22ld (ro)
          /var/tls/cert from cert (ro)
```

> Note: Since AWS IRSA works by mounting the service-account token to pods, the pod spec will be patched and pods recreated if we choose to use rad register to register the roleARN after radius installation.

Reviewer 6: did we not find any workarounds for this? `rad credential register` restarting control plane pods is not ideal

Author: we do not need to restart pods for register command. I will update the doc.

> Note: Since AWS IRSA works by mounting the service-account token to pods, the pod spec will be patched and pods recreated if we choose to use rad register to register the roleARN after radius installation.

Reviewer: I'm scared of this. For Azure we found a solution that doesn't require restarting Radius. What other options do we have here?

Author: Its a document mistake. We dont restart for registering credential, but we have to enable IRSA at the time of install. This means if a user has radius already wants to use AWS IRSA the user will have to reinstall.

```
rad env update qa --aws-account-id <aws-account-id> --aws-region <aws-region> 
```

At this point, we can use aws AssumeRole to manage resources using environment specific account information. rp (for recipes) and ucp will have code changes that make sure of aws assumerole to be able to use the right credential specific to the radius environment for managing the aws resources.  

```
  client := sts.NewFromConfig(cfg)
  assumeRoleArn := retrieveRoleARNCredential() //has logic to get the roleARN stored as credential
  // Assume the new role
  assumeRoleProvider := stscreds.NewAssumeRoleProvider(client, assumeRoleArn)
  // Create a new config with the new credentials
  cfg.Credentials = assumeRoleProvider
  cfg.Region = region
  // Create a new S3 service client
  s3client := s3.NewFromConfig(cfg)
  // Call the ListBuckets function
  output, err := s3client.ListBuckets(context.TODO(), &s3.ListBucketsInput{})
  :

```

### API design

```
@doc("AWS credential kind")
enum AWSCredentialKind {
  @doc("The AWS IRSA credential")
  IRSA,

> }

Reviewer: We're still going to support access keys right?

Author: yes. I captured just the new lines.

```

```
model AwsAccessKeyCredentialProperties extends AwsCredentialProperties {
  @doc("IRSA kind")
  kind: AWSCredentialKind.IRSA;
 
  @doc("The roleARN")
  roleARN: string;

  @doc("The storage properties")
  storage: CredentialStorageProperties;
}
```

> 

Reviewer 8: Is it possible for a user switch between IAM access keys and IRSA or vice-versa? What would that experience be like?

Author: Since the decision to use IRSA is at install time, the switch would need a reinstall. We should handle the case of mismatched credential type and credentials (ex. user enabled irsa but supplied access key) in code as error conditions.

> Based on this new model we have to update API calls. 

Reviewer: Does this result in a breaking change?

Author: No, there are no breaking changes involved here. There are additions.

## User Experience/ CLI Design

> `rad init` and `rad credential` commands should be updated to support AWS IRSA.

Reviewer: What will this look like?

Author: added, thanks to @Reshrahim  for helping me put this section together.

Also, `rad install kubernetes --set` should be able to set a AWS role-arn. 

### non interactive support

> rad install kubernetes --set global.aws.IRSA.enabled=true #default false

Reviewer: It feels like there's still confusion about whether the role arn is configured at installation time or not. L199 says that `rad install kubernetes` needs to be provided the role arn. L202 says it doesn't need the role arn 🙃 

Let's discuss in the meeting, and then please update the doc so it's consistent.

Author: must be clean now. :-)

if enabled, it should print a success/ failure message after updating the pod spec to mount the token. It should also print a information message to the complete of configuration by using rad credential register and rad env update.

> rad credential register aws --type <IRSA> --role <roleARN>

Reviewer 6: to match the pattern we have today with azure, we may want to do `rad credential register aws irsa` and `rad credential register aws access-key`

Author: @willdavsmith , @Reshrahim "access-key" feels a bit long to type. Could we consider `rad credential register aws secret` instead? any other short-forms that convey the meaning is good too.

Author: rad credential register aws irsa and rad credential register aws access-key is good.

> rad credential register aws --type <access-key> --access-key-id <access-key-id> --secret-access-key <secret-access-key>

Reviewer 3: So, we will have IRSA and Access Key types for AWS?

Author: yes.

> rad credential register aws --type <access-key> --access-key-id <access-key-id> --secret-access-key <secret-access-key>

Reviewer: Can you compare this with the design for the azure credentials commands. I'd like to know if this is consistent with the choices we made there.

> rad env update --aws-account-id <aws-account-id> --aws-region <aws-region> 

Reviewer 9: don't we need an environment created first to do a credential register ?

Author: not neccessary.. they can happen in any order.

### interactive rad init support
```
% rad init --full  
Enter an environment name 
>default 

% rad init --full  
Select your cloud provider                     
1. Azure                                  
> 2. AWS                                     
3.[back] 

% rad init --full  

Select the identity for the AWS cloud provider 

> 1. access key                              

Reviewer 9: ```suggestion
1. IAM access key                              
```

> 2. IRSA  

% rad init --full  

Enter the AWS RoleARN for configuring IRSA: <role-arn-radius>

```

Once radius is installed, we use rad env update to update the aws provider config
```

> % rad env update --aws-account-id <aws-account-id> --aws-region <aws-region> 

Reviewer 6: I think rad init should take care of the environment update. we should ask for the account id and region as part of the init flow (I think we do this today)

> % rad env update --aws-account-id <aws-account-id> --aws-region <aws-region> 

Reviewer 3: Can't this be a part of the IRSA setup?

Author: yes, its not new.

> % rad env update --aws-account-id <aws-account-id> --aws-region <aws-region> 

Reviewer: Does this have to be an additional step? I thought we do this for you today in `rad init --full`.

Author: yes, removed this step since its nothing new.

```

**Sample Output:**

n/a

**Sample Recipe Contract:**

n/a

### CLI Design

`rad init --full` will need to be updated to ask the user to choose between AWS access keys and IRSA. rad credential register aws should now support two credential types, and associated parameters.

> We should update the Radius Helm chart to allow the user to enable IRSA during installation with the `rad install kubernets --set global.aws.IRSA.enabled=true` value. This must be passed to ucp and rp charts through values and when the value is set the pod spec should mount the service-account token.

Reviewer 6: nit: for consistency with Azure we can do `global.awsIRSA.enabled=true`

> We should update the Radius Helm chart to allow the user to enable IRSA during installation with the `rad install kubernets --set global.aws.IRSA.enabled=true` value. This must be passed to ucp and rp charts through values and when the value is set the pod spec should mount the service-account token.

Reviewer 3: How is `rad install kubernets --set global.aws.IRSA.enabled=true` going to look for Role ARN?

Author: it only updates the pod spec to mount the service-account token. we set rolearn using rad credential command.

### Multi-tenancy

> Currently, we support only one credential. However, for future, we might want to support multiple credentials potentially across AWS accounts ( say, one for QA team and other for Dev). The code changes for IRSA should be compatible with these changes in future. Since IRSA setup establishes trust between the cluster - oidc provider and AWS, once we support multiple credentials, we should be able to assume different roleARN based on which environment we are deploying from. We just have to establish right trust entities in all roles.  

Reviewer 2: Is this implying that multi-tenancy is out of scope for this design because Radius doesn't support multiple credentials in general today?

Reviewer 5: I think the code implementation for this IRSA support still can take care of future multi-tenancy though it is not supported today. @nithyatsu to more on this.

Author: yes, like @shalabhms  mentions, the proposed IRSA implementation can take care of multi-tenancy in future(when we support multiple credentials)

Author: add a statement "when radius supports multi-tenancy, IRSA will be compatible with it".

### Implementation Details

#### UCP

UCP is responsible for communicating with AWS to deploy AWS resources. 
UCP should be able to assume the role specified in the credential and use this to communicate with AWS since it has the service-token mounted.

#### Bicep

NA

#### Deployment Engine

NA

#### Core RP  / Recipes RP

Terraform provider requires AWS credentials in order to deploy recipes.

> RP should be able to assume the role specified in the credential and use this to communicate with AWS since its pod will have the service-token mounted.

Reviewer 7: I might have misunderstood this so just want to confirm - Currently RP fetches the credentials stored by UCP, were you thinking of switching from that to RP directly fetching the credentials by doing the assuming role instead?

Author: Thank you for bringing this up. I have not looked into recipe / rp implementation of credential handling. There is no reason for RP to change the way it fetches credentials. It would continue to fetch it through UCP, but now, since it gets a role-arn as credential, it should be updated to be able to use that.

> 

Reviewer: I'm glad you thought about recipes 👍

### Error Handling

1. handle mismatch of credential type and arguments supplied.
2. what if the cluster does not have a oidc provider configured?

> 3. what if the user has aws credentials (secrets) configured and wants to switch to IRSA? we can print a informational message requesting user to delete the old credentials since we currently support only one credential. 

Reviewer: I'd like to think this through at a mechanics level. What does the user do to "switch"?

## Test plan

> We should add a E2E to test the deployment of AWS resources using IRSA. 

Reviewer 6: so that we don't have to rotate credentials in the future, IMO we should switch the credentials in our functional tests to using IRSA

Author: makes sense. But then, how can we make sure the access key credential functionality does not break due to some regression?

Author: I think FT + UT we have can help cover it.

> We should add a E2E to test the deployment of AWS resources using IRSA. 

Reviewer: Do we need any new functional tests?

What about testing `rad init`?

Author: rad init in CICD functional test could change to using IRSA like @willdavsmith  said, so that we can test it AND we don't have to worry about rotating credentials. In addition, we should cover the enw credential type in radius/test/functional-portable/ucp/cloud/aws_credential_test.go

Author: add rad init tests where applicable.

## Security

> 

Reviewer: I feel like we should say *something* about security in a design doc for authentication features.

Author: added. :)

## Compatibility

## Monitoring and Logging

We will have the same monitoring and logging as today. We will not be adding any new instrumentation.

> ## Development plan

Reviewer 10: can this work be parallelized?

Author: Yes. Once the model changes are finalized and merged, we should be able to parallelize the efforts (tasks number indicate sequence, same-numbered tasks can be parallelized)
1. model changes get finalized and merged
2. changes to support UCP - AWS resource deployment
2. `rad install` changes to add helm chart configurations
2. `rad credential` changes
3. changes to RP to allow terraform recipes deployment (depends on UCP to fetch credential like @kachawla pointed out)
4. `rad init` changes

we would want to merge all the cli changes in only after the server-side support is complete.

This facilitates a user application to be able to configure its container to use IRSA to authenticate itself for communication with AWS. While the technology is IRSA, this feature is very different from enabling Radius to use IRSA to deploy resources.

### User scenarios

#### User Story 1

As a user I should be able to configure Radius AWS provider with IRSA through interactive experience.

```
% rad init --full  
Enter an environment name 
>default 

% rad init --full  
Select your cloud provider                     
1. Azure                                  
> 2. AWS                                     
3.[back] 

% rad init --full  

Select the identity for the AWS cloud provider 
1. IAM access key                              
> 2. IAM Role for Service Account (IRSA)

% rad init --full  

Enter the IAM RoleARN for configuring IRSA: <role-arn-radius>
```

#### User story 2

As a user I should be able to configure Radius AWS provider with IRSA using non interactive commands.

1. Install the `rad` CLI.

2. Install Radius using `rad install kubernetes`
    ```
    rad install kubernetes --set global.awsIRSA.enabled=true #default false
    ```

>     If enabled, it should print a success/ failure message after updating the pod spec to mount the token. It should also print a information message to the complete of configuration by using rad credential register and rad env update.

Reviewer 6: do we need the roleARN at this step? since we have to update the pod spec and change the service accounts?

Reviewer 6: IMO we shouldn't print any information messages during rad install kubernetes, just update the docs to say what we want them to do next

Author: @Reshrahim and I had a discussion, where we thought we should print a message, since otherwise we have an "incomplete" setup, where user has IRSA enabled and radius expects a roleARN, and the user should use rad credentials to complete the setup. It makes sense for the behavior to be consistent across all credentials though, we can discuss this today.

Author: @willtsai and @Reshrahim  to decide on behavior.

3. Create a Radius environment and add an AWS cloud provider.
    ```
    rad group create default

    rad env create default
    
    rad env update default --aws-account-id <aws-account-id> --aws-region <aws-region> 
    ```

4. Register the role ARN using `rad credential register aws irsa`.
    ```
   rad credential register aws irsa --iam-role <roleARN>
    ```

Radius is now configured to use AWS IRSA for authentication.

## Design

### High Level Design for IRSA

During installation, Radius should allow user to enable IRSA for AWS while conifguring AWS provider. Then the user should be able to enter the roleARN associated with desired policies which will be stored as the AWS credential.

At a high level, the full flow to setup IRSA looks as below:

1. configure the cluster with oidc provider. have the role-arn configured with desired policies ready.

> 2. take steps to allow  `radius-role-arn` and `eks cluster : namespace: service` to trust each other using aws portal.

Reviewer 6: is this possible to do via CLI?

Author: no, since that means radius will need a credential..

Author: document steps using aws cli for feature documentation efforts.

# Radius AWS Workload Identity Support

* **Status**: In Review
* **Author**: Nithya Subramanian (@nithyatsu)

## Overview

A software workload such as a container-based application, service or script needs an identity to authenticate, access, and communicate with services that are distributed across different platforms and/or cloud providers. Radius uses static credentials (Access Key) to interact with AWS IAM (Identity and Access Management) so that we can deploy and access AWS resources today:
 
```
rad credential register aws --access-key-id <access-key-id> --secret-access-key <secret-access-key>
```
 
These IAM credentials, including the access key should be rotated regularly to reduce the chance of unauthorized access. A more secure option than using these credentials is to use IRSA.
 

> IRSA (IAM Roles for Service Accounts) is used for authenticating applications running within Kubernetes pods. When we use IRSA, we associate an IAM role directly with a Kubernetes service-account, allowing the pods associated with the service-account to assume that role. IRSA does not require us to configure access key and  secret. It relies identity tokens (OIDC tokens) to authenticate with AWS services. These tokens have a default expiration (configurable). However, they automatically refresh when needed, transparent to the user.

Reviewer: ```suggestion
IRSA (IAM Roles for Service Accounts) is used for authenticating applications running within Kubernetes pods. When we use IRSA, we associate an IAM role directly with a Kubernetes service-account, allowing the pods associated with the service-account to assume that role. IRSA does not require us to configure access key and secret. It relies upon identity tokens (OIDC tokens) to authenticate with AWS services. These tokens have a default expiration (configurable). However, they automatically refresh when needed and are transparent to the user.
```

Reviewer: Just minor grammar and formatting. The explanation is already good!

The goal of the scenario is to enable infrastructure operators to configure IRSA support for the AWS provider in Radius to deploy and manage AWS resources.

## Terms and definitions
| Term | Definition |
|---|---|
| OIDC | OIDC stands for OpenID Connect. It allows authentication of end-user(or workload)|

> |Role ARN| ARN associated with an AWS IAM role|

Reviewer: nit: These definitions are technically correct, but could explain more to help readers understand the document. 

eg: OIDC is used by IRSA to exchange a Kubernetes token for an AWS IAM token ...

eg: The Role ARN uniquely identifies an AWS IAM Role and can be mapped to a set of permissions. The Role ARN is used to identify the "user" like a ClientID in Azure/Entra or a service account name in Kubernetes.

> |Role ARN| ARN associated with an AWS IAM role|

Reviewer 8: IRSA can manage multiple roles at the same time. Are we allowing/handling that?

Author: the design will allow it, but only after we bring in multiple credential support.

## Objectives

> Issue Reference: https://github.com/radius-project/radius/issues/7618 

### Goals

* Radius users can configure AWS provider and enable Radius to use IRSA for authentication and deployment of application.
* IRSA can be configured via interactive experience.
* IRSA can be configured with non-interactive commands.
* Radius users can deploy and manage AWS resources without having to manage AWS access key id and secret

### Non-goals

* Azure Managed Identity support
* Azure Workload Identity support

> The above two are addressed in [Azure Workload Identitu Support](https://github.com/radius-project/design-notes/blob/main/cli/2024-04-azure-workload-identity.md)

Reviewer 7: nit

```suggestion
The above two are addressed in [Azure Workload Identity Support](https://github.com/radius-project/design-notes/blob/main/cli/2024-04-azure-workload-identity.md)
```

Reviewer 5: +1 . The link seems broken also.

Author: fixed the spelling. The design note is still in PR. I think this would be the link once it is merged cc @willdavsmith . I will check it again at that point.

> The above two are addressed in [Azure Workload Identitu Support](https://github.com/radius-project/design-notes/blob/main/cli/2024-04-azure-workload-identity.md)

Reviewer: ```suggestion
The above two are addressed in [Azure Workload Identity Support](https://github.com/radius-project/design-notes/blob/main/cli/2024-04-azure-workload-identity.md)
```

> * Ability of radified applications to be able to use IRSA for authenticating AWS operations.

Reviewer 9: ```suggestion
* Ability of radified applications to be able to use IRSA to connect to AWS resources.
```

> This facilitates a user application to be able to configure its container to use IRSA to authenticate itself for communication with AWS. While the technology is IRSA, this feature is very different from enabling Radius to use IRSA to deploy resources.

Reviewer: ```suggestion
This facilitates a user application to be able to configure its container to use IRSA to authenticate itself for communication with AWS. While the technology is IRSA, this feature is very different from enabling Radius to use IRSA to deploy resources. We plan to enable this scenario in the future, but it is not part of this proposal. 
```

### User scenarios

#### User Story 1

As a user I should be able to configure Radius AWS provider with IRSA through interactive experience.

```
% rad init --full  
Enter an environment name 
>default 

% rad init --full  
Select your cloud provider                     
1. Azure                                  
> 2. AWS                                     
3.[back] 

% rad init --full  

Select the identity for the AWS cloud provider 
1. IAM access key                              
> 2. IAM Role for Service Account (IRSA)

% rad init --full  

Enter the IAM RoleARN for configuring IRSA: <role-arn-radius>
```

#### User story 2

As a user I should be able to configure Radius AWS provider with IRSA using non interactive commands.

1. Install the `rad` CLI.

2. Install Radius using `rad install kubernetes`
    ```

>     rad install kubernetes --set global.awsIRSA.enabled=true #default false

Reviewer 6: do we need the roleARN at this step? since we have to update the pod spec and change the service accounts?

Author: no.. its the credential instaed.

Author: cover the explanation in alternative design option section on why we did not choose webhook.

    ```
    If enabled, it should print a success/ failure message after updating the pod spec to mount the token. It should also print a information message to the complete of configuration by using rad credential register and rad env update.

3. Create a Radius environment and add an AWS cloud provider.
    ```
    rad group create default

    rad env create default
    

>     rad env update default --aws-account-id <aws-account-id> --aws-region <aws-region> 

Reviewer 5: Side note: What do you think if we can minimize the parameters differentiation in the cli - like keeping just  "--account-id" ,  "--region" and tied to specific aws internally? @nithyatsu , @rynowak

Author: I understand this idea is to make the rad cli command have uniform options, than introduce new ones per cloud instance. will defer to PM and mgmt since it is a different design.

Reviewer: If anyone is interested in this please raise a separate proposal. I don't think it's in scope of this design doc.

    ```

4. Register the role ARN using `rad credential register aws irsa`.
    ```

>    rad credential register aws irsa --iam-role <roleARN>

Reviewer 6: what does rad credential register aws irsa do? we need to either:
1. provide the role-arn at install time, or
2. restart the control plane pods when we run `rad credential register` (probably not good)

Author: clarified in design session.

    ```

Radius is now configured to use AWS IRSA for authentication.

## Design

### High Level Design for IRSA

> During installation, Radius should allow user to enable IRSA for AWS while conifguring AWS provider. Then the user should be able to enter the roleARN associated with desired policies which will be stored as the AWS credential.

Reviewer: It might be better stated as "users must choose whether or not to enable IRSA during Radius installation".

At a high level, the full flow to setup IRSA looks as below:

1. configure the cluster with oidc provider. have the role-arn configured with desired policies ready.
2. take steps to allow  `radius-role-arn` and `eks cluster : namespace: service` to trust each other using aws portal.
3. use [interactive](#user-story-1) approach or [non-interative](#user-story-2) to configure Radius to use IRSA
4. during rad deploy, radius assumes the role that is configured in the credential. 

OIDC is configured for a cluster and not specific to a role-arn. Therefore, once we have support for multiple crdentials in Radius, the same solution would work for supporting multitenancy.

More details about each step is covered in [Detailed Design](#detailed-design)

IRSA works the same way on EKS as well as non-EKS clusters.  

### Architecture Diagram

## User Experience/ CLI Design

Following commands should be updated for AWS IRSA support

`rad init` should be updated to allow configuring a AWS IRSA provider.

`rad credential` command should be updated to register IAM RoleARN.

> `rad install kubernetes --set` should be able to enable IRSA. 

Reviewer: Does this imply that we're updating the Helm chart?

Reviewer: Can you update the doc to make this clear? We're not adding any new logic to `rad install kubernetes`. It's going in the helm chart 👍

**Sample Output:**

n/a

**Sample Recipe Contract:**

n/a

### Detailed Design

#### Proposed Option

Radius should support AWS IRSA.

#### Prerequisite

1. [Setup OIDC provider] (https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html) should be complete for the cluster. 
   

> 2. `radius-role` with desired policies should be created. Its trust relation should look similar to below. This establishes a trust relation between the annotated service-accounts in the specified namespace in specified cluster with the specified OIDC provider to be established. 

Reviewer: Which service account(s) need to be registered as trusted for the role?

Author: ucp and rp ( updated doc)

  ```
  {
    "Version": "2012-10-17",
    "Statement": [
      {
        "Effect": "Allow",
        "Principal": {
          "Federated": "arn:aws:iam::${ACCOUNT_ID}:oidc-provider/${OIDC_PROVIDER}"
        },
        "Action": "sts:AssumeRoleWithWebIdentity",
        "Condition": {
          "StringEquals": {

>             "${OIDC_PROVIDER}:aud": "<AUDIANCE-NAME>",

Reviewer 7: For audience type, is anything else apart from `sts.amazonaws.com` supported?

```suggestion
            "${OIDC_PROVIDER}:aud": "<AUDIENCE-NAME>",
```

Author: need not templatize sts.

            "${OIDC_PROVIDER}:sub": "system:serviceaccount:<NAMESPACE>:<SERVICE-ACCOUNT>"
          }
        }
      }
    ]
  }
  ```
Example:
  ```
  {
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Federated": "arn:aws:iam::817312594854:oidc-provider/oidc.eks.us-west-2.amazonaws.com/id/67DDAC18D8C44CEDCF1C9719A8E9B866"
            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "StringEquals": {

>                     "oidc.eks.us-west-2.amazonaws.com/id/67DDAC18D8C44CEDCF1C9719A8E9B866:sub": "system:serviceaccount:radius-system:ucp",

Reviewer 7: Would we need multiple policies for ucp, rp, de?

Author: yes, each pf ucp, rp need the configuration.

                    "oidc.eks.us-west-2.amazonaws.com/id/67DDAC18D8C44CEDCF1C9719A8E9B866:aud": "sts.amazonaws.com"
                }
            }
        },
    ]
}
  ```

the user-specific details would be {ACCOUNT_ID} and {OIDC_PROVIDER}.

#### Radius support

**installation**

> User can enable IRSA for AWS by by choosing "IRSA" option by choosing IRSA during rad init and providing a RoleARN (`arn:aws:iam::123456789012:role/radius-role`). This stores the roleARN as credential. It also mounts the service-account token to ucp and rp pods as secret.

Reviewer: I think it's better if the detailed design covers the approach step-by-step rather than by discussing `rad init`. `rad init` is an automation layer that simplifies what is possible through the API 😉.

We should answer: 

> What steps need to be taken in Radius, and what information does the user provide at each step?

> User can enable IRSA for AWS by by choosing "IRSA" option by choosing IRSA during rad init and providing a RoleARN (`arn:aws:iam::123456789012:role/radius-role`). This stores the roleARN as credential. It also mounts the service-account token to ucp and rp pods as secret.

Reviewer 7: > mounts the service-account token to ucp and rp pods as secret

Which namespace does this secret belong to?

Author: it is in the radius-system namespace.

> User can enable IRSA for AWS by by choosing "IRSA" option by choosing IRSA during rad init and providing a RoleARN (`arn:aws:iam::123456789012:role/radius-role`). This stores the roleARN as credential. It also mounts the service-account token to ucp and rp pods as secret.

Reviewer 8: Maybe I'm confusing something here but I saw above that user does not need to provide roleARN during installation.

Author: the confusion is because rad init also does a rad install, if radius isn't installed. I will rewrite the section in terms of individual commands to improve clarity.

```
    Containers:
      ucp:
        Container ID:   docker://faa5af929f91e053b04b023e05c4fa6363e87688df265443d9f377258fc3fd04
        Image:          ghcr.io/radius-project/ucpd:latest
        Image ID:       docker-pullable://ghcr.io/radius-project/ucpd@sha256:991ed703a3260a19535690242cef4d7e4d076038b016443d9106f2d76d57b0da
        Ports:          9443/TCP, 9090/TCP
        Host Ports:     0/TCP, 0/TCP
        :
        Environment:
          UCP_CONFIG:                   /etc/config/ucp-config.yaml
          BASE_PATH:                    /apis/api.ucp.dev/v1alpha3
          TLS_CERT_DIR:                 /var/tls/cert
          PORT:                         9443
        Mounts:
          /etc/config from config-volume (rw)
          /var/run/secrets/eks.amazonaws.com/serviceaccount from aws-iam-token (ro)
          /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-d22ld (ro)
          /var/tls/cert from cert (ro)
```

> Note: Since AWS IRSA works by mounting the service-account token to pods, the decision to use IRSA should be taken at install time. Credentials can be registered later on.

Author: update the instructions in terms of individual rad commands than rad init.

**post installation**

Once the installation is complete, the user can use rad env update to store relevant information for deploying resources to accounts and region associated with specific radius environment.

```
rad env update qa --aws-account-id <aws-account-id> --aws-region <aws-region> 
```

> At this point, we can use aws AssumeRole to manage resources using environment specific account information. rp (for recipes) and ucp will have code changes that make sure of aws assumerole to be able to use the right credential specific to the radius environment for managing the aws resources.  

Reviewer 7: I missed clarifying this before, will we update DE to call assumeRole and fetch the credentials on the fly or would it call into ucp to get the credentials?

Reviewer 7: never mind, DE doesn't call into Azure, it justs route the request back to UCP :)

```
  client := sts.NewFromConfig(cfg)
  assumeRoleArn := retrieveRoleARNCredential() //has logic to get the roleARN stored as credential
  // Assume the new role
  assumeRoleProvider := stscreds.NewAssumeRoleProvider(client, assumeRoleArn)
  // Create a new config with the new credentials
  cfg.Credentials = assumeRoleProvider
  cfg.Region = region
  // Create a new S3 service client
  s3client := s3.NewFromConfig(cfg)
  // Call the ListBuckets function

>   output, err := s3client.ListBuckets(context.TODO(), &s3.ListBucketsInput{})

Reviewer: I think it's really good that you found a code sample and got it working. We should take the explanation further.

Can you update this to include some more details about what needs to change in Radius? How do we get AWS credentials today? Where in the code/system? What will be different when we add IRSA support?

Reviewer: Thanks for adding this 👍

  :

```

### API design

```
@doc("AWS credential kind")
enum AWSCredentialKind {
  @doc("The AWS Access Key credential")
  AccessKey,

  @doc("The AWS IRSA credential")
  IRSA,
}
```

```

> model AwsIRSACredentialProperties extends AwsCredentialProperties {

Reviewer: We should be consistent about captialization of initialisms. I see `AWS` (go style) on line 261 and `Aws` (javascript style) on line 272.

Reviewer: What's the plan for this?

  @doc("Access Key kind")
  kind: AWSCredentialKind.IRSA;

  @doc("RoleARN for AWS IRSA identity")

>   @secret

Author: not a secret.

>   @secret

Reviewer: This is not a secret.

  roleARN: string;

>   @doc("The storage properties")

Reviewer 7: Question about credentials management in general - what do we do with this property?

Author: @rynowak , could you please help me understand this as well.

Reviewer: Storage is used to determine how we store the secret data. Right now the only option is "internal". This was part of the TLS design, and we wanted to leave options for the future to support other options.

eg: my data is in KeyVault.

  storage: CredentialStorageProperties;
}
```

> Based on this new model we have to update API calls. 

Reviewer: Can you say more?

Author: rephrased for correctness.

### CLI Design

* `rad init --full` should be updated to ask the user to choose between AWS access keys and IRSA. If IRSA is chosen, the user should be able to input a AWS IAM Role ARN. As a result of rad init, 
   1. the helm charts should be updated with service-account token mounted as secret.

>    2. roleARN should be stored as a AWS credential

Reviewer 7: > roleARN should be stored as a AWS credential

a bit confused by this statement.. what do we mean by roleARN as AWS credential?

Author: roleARN would be stored as the radius credential for AWS. Radius would assume this role in order to communicate with AWS. This would work since the cluster is configured with OIDC provider ,pods would have service-account token mounted, and role has the right trust entries to trust the radius service-accounts.

* `rad install kubernetes --set global.aws.IRSA.enabled=true` should add the neccessary pod spec to mount the service-account token.

* `rad credential register aws --access-key-id <access-key-id> --secret-access-key <secret-access-key>` should be replaced with 

> `rad credential register aws secret --access-key-id <access-key-id> --secret-access-key <secret-access-key>`

Reviewer: ```suggestion
`rad credential register aws secret --access-key-id <access-key-id> --secret-access-key <secret-access-key>`
```

Is `secret` the right noun?

Author: fixed to use access-key as decided in the meeting.

> `rad credential register aws secret --access-key-id <access-key-id> --secret-access-key <secret-access-key>`

Reviewer 7: This seems inconsistent from the interactive experience above where this option is named as "IAM access key". Any reason to name it as "secret" here

### Implementation Details

#### UCP

UCP is responsible for communicating with AWS to deploy AWS resources. 
UCP should be able to assume the role specified in the credential and use this to communicate with AWS since it has the service-token mounted.

#### Bicep

NA

#### Deployment Engine

NA

#### Core RP  / Recipes RP

Terraform provider requires AWS credentials in order to deploy recipes.

RP should be able to assume the role specified in the credential and use this to communicate with AWS since its pod will have the service-token mounted.

### Error Handling

1. handle mismatch of credential type and arguments supplied.
2. what if the cluster does not have a oidc provider configured?
3. what if the user has aws credentials (secrets) configured and wants to switch to IRSA? 
   1. the user has to re-install radius to disable IRSA. This should print informational messages to unregister IRSA credential and re-register the access key credentials.
   2. If there is a mismatch (pod spec has IRSA enabled, but credential type is not role-arn or vice-versa, deployment will fail. Enhance error message to suggest checking the right kind of credential is configured)

> 

Author: error - detectable/ not, if we can detect, help user to fix it, otherwise, point a documentation that can help troubleshoot.

>    2. If there is a mismatch (pod spec has IRSA enabled, but credential type is not role-arn or vice-versa, deployment will fail. Enhance error message to suggest checking the right kind of credential is configured)

Reviewer: Are these open questions? Or things you've figured out?

Author: Just some points I had noted to cover during implementation. No open questions.

## Test plan

> We should add a E2E to test the deployment of AWS resources using IRSA. 

Author: revisit once we have multiple credential support (log a ticket)

> We should add a E2E to test the deployment of AWS resources using IRSA. 

Author: explore the close to E2E strategy for this feature without a lot of overhead + new pipeline.

## Security

AWS IRSA does not rely of user to rotate credentials for increased security. 
Thus, it should be preferred credential type. User should configure the role associated with Radius setup following principle of least privilege, so that Radius has just the sufficient policies for managing the supported AWS resources.

## Compatibility
NA

> 

Reviewer: This is a breaking change to the CLI. Is there any code outside of the Radius repo that needs to be updated?

Author: thank you for calling this out. The documentation should be updated too.  cli docs are auto generated but we have to review the samples to make sure the right command is used.

## Monitoring and Logging

We will have the same monitoring and logging as today. We will not be adding any new instrumentation.

## Development plan

* Create POC for Radius + AWS (1 engineer,1 sprint)  [complete] 
* Create and review technical design (1 engineer, 0.5 sprint) [in progress]
* Implement model changes (1 engineer, 0.5 sprint) 
* Implement changes in UCP, and Recipes RP (1 engineer, 1 sprint)
* Implement CLI and Helm chart changes (1 engineer, 1 sprint) 
* End-to-end testing and documentation (1 engineer, 1 sprint) 

## Open Questions

> ## Alternatives considered 

Reviewer: Either alternatives or design review notes would be a great place to discuss the topic of multiple roles + multi-tenancy. We're ignoring some of the built-in features of IRSA (the webhook) because our scenario is different.

 
IRSA (IAM Roles for Service Accounts) is used for authenticating applications running within Kubernetes pods. When we use IRSA, we associate an IAM role directly with a Kubernetes service-account, allowing the pods associated with the service-account to assume that role. IRSA does not require us to configure access key and  secret. It relies identity tokens (OIDC tokens) to authenticate with AWS services. These tokens have a default expiration (configurable). However, they automatically refresh when needed, transparent to the user.

The goal of the scenario is to enable infrastructure operators to configure IRSA support for the AWS provider in Radius to deploy and manage AWS resources.

## Terms and definitions

## Objectives

> Issue Reference: https://github.com/radius-project/radius/issues/7618 

### Goals

* Radius users can configure AWS provider and enable Radius to use IRSA for authentication and deployment of application.
* IRSA can be configured via interactive experience.
* IRSA can be configured manually.
* Radius users can deploy and manage AWS resources without having to manage AWS access key id and secret

> ### Non-goals

Reviewer 9: should we call out the IRSA support for applications to connect to AWS support is separate and is a non-goal to this item ?

Reviewer: Yes please :)

Author: I will update the non goal so its clear and add the reason.

> * Azure Managed Identity support

Reviewer 9: This might be irrelevant to this document?

Author: Add the link to document that covers this.

Author: work with @Reshrahim on the text.

* Azure Workload Identity support
* Ability of radified applications to be able to use IRSA for authenticating AWS operations.

### User scenarios

> As a user, I should be able to configure IRSA with Radius for managing AWS resources.

Reviewer 9: Should we have the interactive and manual experience called out here as the two user scenarios with step by step instructions?

Author: updated document to have the user scenarios in the beginning.

## Design

### High Level Design for IRSA

During installation, user can enable IRSA for AWS by by choosing "IRSA" option while configuring AWS provider and providing the Role ARN. Then, service-accounts will be annotated with role arn, similar to `amazonaws.com/role-arn: arn:aws:iam::123456789012:role/rad-role`. This allows the service-account to assume the specified role. A trust relation between the annotated service-accounts in the specified namespace with the specified OIDC provider is established this way.  

[Amazon EKS Pod Identity Webhook](https://github.com/aws/amazon-eks-pod-identity-webhook#amazon-eks-pod-identity-webhook) installed in cluster can inject neccessary configurations into the pods associated with these service-accounts for IRSA. 

At this point, we can use aws-sdk to work with IRSA to deploy and manage AWS resources. Specifically, AWS assume-role helps handle multi-tenancy. At a high level, the flow looks as below:

1. configure the cluster with oidc provider and create a role ex: `rad-role` with no policies.
2. install eks pod identity webhook.
3. use  `rad init` to configure radius with role-arn associated with IRSA (`rad-role-arn`). 

> 4. use `rad env update` to configure role-arn associated with the specific environment (`rad env update qa --aws-account-id <aws-account-id> --aws-region <aws-region> --aws-role-arn <qa-role-arn>`). This is the role to be used when user deploys from `qa` env. Should be configured with appropriate policies.

Author: updated document, this was a doc mistake.

   rad credential register aws irsa --iam-role <roleARN>
    ```

Radius is now configured to use AWS IRSA for authentication.

## Design

### High Level Design for IRSA

Users must choose whether or not to enable IRSA during Radius installation. Then the user should be able to enter the roleARN associated with desired policies which will be stored as the AWS credential.

At a high level, the full flow to setup IRSA looks as below:

1. configure the cluster with oidc provider. 
2. have the role-arn configured with desired policies ready.
3. take steps to allow  `radius-role-arn` and `eks cluster : namespace: service` to trust each other using aws portal.
4. use [interactive](#user-story-1) approach or [non-interative](#user-story-2) to configure Radius to use IRSA.
5. during rad deploy, radius assumes the role that is configured in the credential. 

OIDC is configured for a cluster and not specific to a role-arn. Therefore, once we have support for multiple crdentials in Radius, the same solution would work for supporting multitenancy.

More details about each step is covered in [Detailed Design](#detailed-design)

IRSA works the same way on EKS as well as non-EKS clusters.  

> ### Architecture Diagram

Reviewer 7: Is there something we put together for this to show the call flow between components and to AWS?

`rad credential` command should be updated to register IAM RoleARN.

`rad install kubernetes --set` should be able to enable IRSA. 

**Sample Output:**

n/a

**Sample Recipe Contract:**

n/a

### Detailed Design

#### Proposed Option

Radius should support AWS IRSA.

#### Prerequisite

1. [Setup OIDC provider] (https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html) should be complete for the cluster. 
   
2. `radius-role` with desired policies should be created. Its trust relation should look similar to below. This establishes a trust relation between the service-accounts (ucp and applications-rp) in the specified namespace (radius-system) in specified cluster with the specified OIDC provider to be established. 

  ```
  {
    "Version": "2012-10-17",
    "Statement": [
      {
        "Effect": "Allow",
        "Principal": {
          "Federated": "arn:aws:iam::${ACCOUNT_ID}:oidc-provider/${OIDC_PROVIDER}"
        },
        "Action": "sts:AssumeRoleWithWebIdentity",
        "Condition": {
          "StringEquals": {
            "${OIDC_PROVIDER}:aud": "sts.amazonaws.com",

>             "${OIDC_PROVIDER}:sub": "system:serviceaccount:<NAMESPACE>:<SERVICE-ACCOUNT>"

Reviewer 7: Will the namespace always be `radius-system`?

Author: add a descriotion to namespace.

                "StringEquals": {
                    "oidc.eks.us-west-2.amazonaws.com/id/67DDAC18D8C44CEDCF1C9719A8E9B866:sub": "system:serviceaccount:radius-system:ucp",
                    "oidc.eks.us-west-2.amazonaws.com/id/67DDAC18D8C44CEDCF1C9719A8E9B866:aud": "sts.amazonaws.com"
                }
            }
        },
    ]
}
  ```

  ```
  {
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Federated": "arn:aws:iam::817312594854:oidc-provider/oidc.eks.us-west-2.amazonaws.com/id/67DDAC18D8C44CEDCF1C9719A8E9B866"
            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "StringEquals": {
                    "oidc.eks.us-west-2.amazonaws.com/id/67DDAC18D8C44CEDCF1C9719A8E9B866:sub": "system:serviceaccount:radius-system:applications-rp",
                    "oidc.eks.us-west-2.amazonaws.com/id/67DDAC18D8C44CEDCF1C9719A8E9B866:aud": "sts.amazonaws.com"
                }
            }
        },
    ]
}
  ```

the user-specific details would be {ACCOUNT_ID} and {OIDC_PROVIDER}.

#### Radius support

**installation and setup**

> * use `rad install kubernetes --set global.aws.IRSA.enabled=true` to add the neccessary pod spec to mount the service-account token.

Reviewer: nit: it feels more consistent to call this setting `global.aws.irsa.enabled`. 

AWS is an initialism and IRSA is an acronym. We should either put both in caps or put both in lowercase.

        Host Ports:     0/TCP, 0/TCP
        :
        Environment:
          UCP_CONFIG:                   /etc/config/ucp-config.yaml
          BASE_PATH:                    /apis/api.ucp.dev/v1alpha3
          TLS_CERT_DIR:                 /var/tls/cert
          PORT:                         9443
        Mounts:
          /etc/config from config-volume (rw)
          /var/run/secrets/eks.amazonaws.com/serviceaccount from aws-iam-token (ro)
          /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-d22ld (ro)
          /var/tls/cert from cert (ro)
```

* use `rad credential register aws irsa --iam-role <roleARN>` to register the roleARN that radius would assume for deploying and managing AWS resources.
This command "PUT"s a Radius Credential Resource. As part of this, the credential gets stored as a kubernetes secret (similar to how Radius manages all credentials today).

* use rad env update to store relevant information for deploying resources to accounts and region associated with specific radius environment. (This step is not newly introduced)

```
rad env update qa --aws-account-id <aws-account-id> --aws-region <aws-region> 
```

**post installation**

> Once the steps in instalation and setup is completed, UCP should be able to "fetch" the configured credentials and then use AssumeRole to manage AWS resources. AWSCredentialProvider should be updated to support the new credential. This can then be used as part of initializing the UCP AWS Module with IRSA credentials. RP should fetch the credential through UCP and utilize it with Terraform provider.  

Reviewer 7: nit

```suggestion
Once the steps in installation and setup is completed, UCP should be able to "fetch" the configured credentials and then use AssumeRole to manage AWS resources. AWSCredentialProvider should be updated to support the new credential. This can then be used as part of initializing the UCP AWS Module with IRSA credentials. RP should fetch the credential through UCP and utilize it with Terraform provider.  
```

> Once the steps in instalation and setup is completed, UCP should be able to "fetch" the configured credentials and then use AssumeRole to manage AWS resources. AWSCredentialProvider should be updated to support the new credential. This can then be used as part of initializing the UCP AWS Module with IRSA credentials. RP should fetch the credential through UCP and utilize it with Terraform provider.  

Reviewer 7: For `RP should fetch the credential through UCP and utilize it with Terraform provider.` - do we need to set any additional property in the Terraform config?

Author: open question: cover any terraform config that should be added as part of design.

Author: added

    TokenExpirationSeconds:  86400

If we chose to use the web-hook , we could have annotated the ucp and rp service-accounts with roleARN and installed web-hook. However, we did not choose this approach because -
* it requires roleARN to be injected at install time. Dealing with credentials should not require a service restart. 
* when Radius starts supporting multi-tenancy, there is no support in the webhook to handle annotation of service-accounts with multiple role-arns.  

### EKS Pod Identity 

EKS Pod Identity was introduced in 2022 as a simplified approach for applications running on EKS to retrieve credentials. It uses the new EKS pod identity instead of OIDC provider for identity. Because of this, users have convenient APIs that allow managing pod identity. 
Ref. [AWS EKS Pod Identity](https://aws.amazon.com/blogs/containers/amazon-eks-pod-identity-a-new-way-for-applications-on-eks-to-obtain-iam-credentials/). 

Its good to note that since this approach uses its own service principal, the setup involves different identity providers and configuration on aws. Radius should be however able to support this feature with minimal code changes, if any.  

> 

Reviewer: Thanks for adding these 👍

## Design Review Notes

<!--
Update this section with the decisions made during the design review meeting. This should be updated before the design is merged.

> -->

Reviewer: Anything that should go here?

Author: add "see alternatives"
